packages/k2-alpine/src/components/PinInput/PinInput.tsx

```diff
@@ -0,0 +1,232 @@
+import React, { forwardRef, useImperativeHandle, useMemo, useRef } from 'react'
+import { Platform, TextInput, Vibration, ViewStyle } from 'react-native'
+import Animated, {
+  cancelAnimation,
+  runOnJS,
+  SharedValue,
+  useAnimatedStyle,
+  useSharedValue,
+  withDelay,
+  withSequence,
+  withSpring,
+  withTiming
+} from 'react-native-reanimated'
+import { TouchableOpacity } from '../Primitives'
+import { useTheme } from '../..'
+
+export const PinInput = forwardRef<PinInputActions, PinInputProps>(
+  ({ value, onChangePin, length = 6, style }, ref) => {
+    const textInputRef = useRef<TextInput>(null)
+    const wrongPinAnimation = useSharedValue(0)
+    const loadingDotAnimations = useLoadingDotAnimations(length)
+    const isLoading = useSharedValue(false)
+
+    const startLoadingAnimation = (): void => {
+      const triggerAnimations = async (): Promise<void> => {
+        const animationPromises = loadingDotAnimations.map(
+          (sharedValue, index) =>
+            new Promise<void>(resolve => {
+              sharedValue.value = withSequence(
+                withDelay(index * 50, withTiming(-15, { duration: 200 })),
+                withTiming(0, { duration: 200 })
+              )
+
+              setTimeout(() => {
+                runOnJS(resolve)()
+              }, index * 50 + 400)
+            })
+        )
+
+        return Promise.all(animationPromises).then(() => {
+          if (isLoading.value) {
+            runOnJS(triggerAnimations)()
+          }
+        })
+      }
+
+      isLoading.value = true
+      triggerAnimations()
```

Comment:
should the first run be enclosed in a runOnJS too?

Reply:
The first call to triggerAnimations() doesn’t need to be enclosed in runOnJS since it’s already running in the JS thread. However, recursive calls inside the animation logic should use runOnJS as they are triggered from the native thread.

Reply:
but I just wrapped it with `runOnJS` for consistency.